onnxruntime/core/providers/cpu/math/top_k.cc

```diff
@@ -22,115 +22,179 @@
+          auto col_index = l * block_slice + j;
+          values_map(i, col_index) = elem.first;
+          indices_map(i, col_index) = elem.second;
+        }
       }
     }
   }
+}
+
+// Wrapper over core TopK implementation
+Status TopKImpl(OpKernelContext* p_op_kernel_context, const Tensor* input, const int axis, const unsigned k,
+                bool largest = true, bool sorted = true) {
+  const TensorShape& input_shape = input->Shape();
+  // Will return axis_ as is if positive or fixes it in case it is negative
+  const auto axis_parsed = HandleNegativeAxis(axis, static_cast<int64_t>(input_shape.NumDimensions()));
+  // Check to ensure k is within the bounds of what is available in that specific axis
+  if (input_shape[axis_parsed] < k) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "k argument [" , k , "] should not be greater than specified axis dim value [" , input_shape[axis_parsed] , "]");
+  }
+
+  // Resize output tensors to be the same shape as the input except
+  // for the specified dimension ((i.e.) axis_parsed), which will be of size k. E.x. for an input tensor
+  // of shape [3, 4, 5] and k=2 with axis_parsed=1, both of the outputs will be shape [3, 2, 5]
+  TensorShape output_shape = input_shape;
+  output_shape[axis_parsed] = k;
+  auto* values = p_op_kernel_context->Output(0, output_shape);
+  auto* indices = p_op_kernel_context->Output(1, output_shape);
+
+  if (values == nullptr || indices == nullptr) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL,
+                  "output count mismatch, expected 2 outputs to be present for TopK operator");
+  }
+
+  // no-op - no output buffers to fill - return silently
+  if (k == 0) {
+    return Status::OK();
+  }
+
+  // extract sorted largest TopK elements
+  if (sorted && largest) {
+    // Average - O(NlogK). Worst - O(NlogK), where 'N' is the number of input elements
+    extract_top_k_elements<true, true, GreaterValueCmp<float>>(input, input_shape, values, indices, output_shape, k, gsl::narrow_cast<unsigned>(axis_parsed));
+  }
+
+  // extract sorted smallest TopK elements
+  else if (sorted && !largest) {
+    // Average - O(NlogK). Worst - O(NlogK), where 'N' is the number of input elements
+    extract_top_k_elements<false, true, LesserValueCmp<float>>(input, input_shape, values, indices, output_shape, k, gsl::narrow_cast<unsigned>(axis_parsed));
+  }
+
+  // extract unsorted (order undefined) largest TopK elements
+  else if (largest) {
+    // Average - O(N). Worst - O(NlogN)/O(N^2) depending on underlying implementation, where 'N' is the number of input elements
+    extract_top_k_elements<true, false, GreaterValueCmp<float>>(input, input_shape, values, indices, output_shape, k, gsl::narrow_cast<unsigned>(axis_parsed));
+  }
+
+  // extract unsorted (order undefined) smallest TopK elements
+  else if (!largest) {
+    // Average - O(N). Worst - O(NlogN)/O(N^2) depending on underlying implementation, where 'N' is the number of input elements
```

Comment:
having 'else if (!largest)' is slightly confusing as really it's just an 'else'

Reply:
Fixed.